_posts/2023-02-03_medical-imaging-with-nix/medical-imaging-with-nix.Rmd

```diff
@@ -0,0 +1,211 @@
+some of my woes but I haven't tried it yet.
+
+Another advantage alluded to already is speed. Compiling things takes time, so unless you're always getting prebuilt binaries from the
+nix servers you can be in for long build times. Nix is good at not duplicating build work, but sometimes rebuilding is unavoidable. Say
+you want to use a newer version of cuda, or gcc, most of your environment will need to be rebuilt^[ a colleague suggests that we create a build
+farm where a background process builds new versions of things so that there are prebuilt versions always available, but that is an infrastructure
+investment that hasn't felt worth it yet. Apparently the issue is that CUDA is technically not free software and by policy the nix
+build farms won't build it for us. ]
+
+Finally, the last difficulty is with irreducible system dependencies. Nixpkgs, when you're not using the nix operating system, does
+not interact well with graphics drivers, there is a wrapper project I use called [NixGL](https://github.com/guibou/nixGL) which gives 
+you access to graphics drivers and allows you to run programs that use CUDA^[you need to start programs with `nixGL<driver> <program>` which
+is irritating, but I think can be fixed in my flake]. However this has meant I needed to hardcode my graphics
+driver into the nix flake, severely hindering portability^[Ok you'd need to edit one line in the flake but that's enough of a barrier
+to discourage users. If anyone in the nix community can help me solve this problem I'd be extremely grateful]. 
+
+## Is this worth it?
+
+I don't know. I've sunk considerable cost into building this environment so one might reasonably expect me to be quite biased. 
+
+Would cutting down my effort by 75% at the risk of some amount of computational irreproducibility and unportability be worth it?
+
+Is this a case of good enough practices trumping best practices? 
+
+I'm not sure, but now that it's built I know that I will be able to revisit this exact environment for many years to come. I've acquired the
+skills to fix package sets as issues arise, so it's not a large burden any longer to develop. If I can iron out a few more details about
+merging python package sets from multiple versions of nixpkgs I would think that this is substantively superior to managing my software
+with conda, apt, renv, and docker.
+
+Would I encourage others to adopt the strategy of creating a nix environment from scratch for each new project? Probably not, at least don't
+go it alone. Sharing working flakes and overlays for data scientists to springboard off of makes this type of reproducible, portable, futuristic
+software versioning possible^[and of course contributing back to nixpkgs where you can], but keep your [novelty budget](https://shimweasel.com/2018/08/25/novelty-budgets) in mind, you might be wise to
+[pick boring technology](https://boringtechnology.club/). 
+
```

Comment:
suggestion: perhaps you could add a quick note about any other future steps you might take with the nix environment and when you use/may use it

Reply:
Added a section to this effect at the bottom, thanks for the suggestion!